### PR TITLE
Refactor error system

### LINKED ISSUE #164: Refactor error system
Currently, most of our defined errors are unused (they were previously created for `hermes` needs). This [creates confusion](https://github.com/cosmos/ibc-rs/pull/161#discussion_r992475283) in how to use the errors properly.

We should also revisit whether we need `flex_error`, or if instead we want to move to a more canonical & simple error crate such as [anyhow](https://docs.rs/anyhow/latest/anyhow/).

### PR DESCRIPTION
**Pull request by DaviRain-Su**
Closes: #164 

## Description

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/ibc/src/applications/transfer/error.rs [hunk 1]
```diff
@@ -15,134 +15,104 @@ use crate::core::ics24_host::identifier::{ChannelId, PortId};
...
-
-        Utf8Decode
-            [ TraceError<Utf8Error> ]
-            | _ | { "error decoding raw bytes as UTF8 string" },
-    }
+#[cfg(feature = "std")]
+impl std::error::Error for Error {}
+
+#[derive(Display, Debug)]
+pub enum Error {
```
Comment (plafer) [inline crates/ibc/src/applications/transfer/error.rs]: Hey thank you for starting to work on this! Could you share what changes you were planning on making? Specifically, it' not clear to me yet what the error system should look like. 

However, I was imagining we'd want to have drastically fewer error variants than this. We might even want to have all our errors just be a newtype around a `String`?

cc @hu55a1n1

Comment (DaviRain-Su) [reply crates/ibc/src/applications/transfer/error.rs]: > However, I was imagining we'd want to have drastically fewer error variants than this. We might even want to have all our errors just be a newtype around a String?

I understand what you mean, for example, the displaydoc used here adds a similar comment to each variant that is the output of each display of the error variant, which is a disguised implementation of the to_string() method.

**Comment by DaviRain-Su** [comment]
Do you have a better description of the error?
```rust
    /// Ics04 channel error(`{0}`)
    Ics04Channel(channel_error::Error),
    /// invalid port identifier `{context}` error(`{validation_error}`)
    InvalidPortId {
        context: String,
        validation_error: ValidationError,
    },
    /// invalid channel identifier `{context}` error(`{validation_error}`)
    InvalidChannelId {
        context: String,
        validation_error: ValidationError,
    },

```

File: crates/ibc/src/applications/transfer/error.rs (hunk 1 above)
Comment (hu55a1n1) [reply crates/ibc/src/applications/transfer/error.rs]: I think we should try to restrict the error variants to those that are relevant to (or created by) this module and all other errors should be wrapped in a module variant -> e.g. `ChannelError(channel_error::Error)`. These errors can be returned as `Error::source()` in the `Error` trait impl as suggested in the [stdlib](https://doc.rust-lang.org/beta/core/error/trait.Error.html) ->
> If one module must report an error that is caused by an error from a lower-level module, it can allow accessing that error via [Error::source()](https://doc.rust-lang.org/beta/core/error/trait.Error.html#method.source). This makes it possible for the high-level module to provide its own errors while also revealing some of the implementation for debugging.

File: crates/ibc/src/applications/transfer/error.rs [hunk 2]
```diff
@@ -15,134 +15,104 @@ use crate::core::ics24_host::identifier::{ChannelId, PortId};
...
-        InvalidCoin
-            { coin: String }
-            | e | { format_args!("invalid coin string: {}", e.coin) },
-
-        Utf8Decode
-            [ TraceError<Utf8Error> ]
-            | _ | { "error decoding raw bytes as UTF8 string" },
-    }
+#[cfg(feature = "std")]
+impl std::error::Error for Error {}
```
Comment (hu55a1n1) [inline crates/ibc/src/applications/transfer/error.rs]: I think it would be nice to implement this. Once `core::error::Error` is stabilized we can rename `std::error::Error` to `core::error::Error` while the implementation remains the same -> 
```rust
#[cfg(feature = "std")]
impl std::error::Error for Error {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            Error::Ics04Channel(e) => Some(e),
            _ => None // TODO
        }
    }
}
```

File: crates/ibc/src/core/ics04_channel/error.rs [hunk 3]
```diff
@@ -12,354 +12,158 @@ use crate::signer::SignerError;
...
-            { description: String }
-            | e| { format_args!("other error: {0}", e.description) },
-    }
-}
-
-impl Error {
-    pub fn chan_open_confirm_proof_verification(e: Error) -> Error {
-        e.add_trace(&"Handshake proof verification fails at ChannelOpenConfirm")
-    }
+#[derive(Debug, Display)]
```
Comment (hu55a1n1) [inline crates/ibc/src/core/ics04_channel/error.rs:17]: I think we're missing the `Error` trait impl  here ->
```rust
#[cfg(feature = "std")]
impl std::error::Error for Error {}
```

**Comment by plafer** [comment]
Note: hopefully our solution here also addresses #23. I'm still thinking about how to make everything work together (logs discarded on error issue, no_std support, error sources, small number of `Error` variants if possible, etc)

EDIT: Actually #23 fixed by the new `ValidationContext`: logs are no longer returned, but rather given to the context with `log_message()`.

**Comment by plafer** [comment]
Here's my current understanding:

+ methods in `ValidationContext` and `ExecutionContext` that are implemented by the host should return either some newtype around `String` if we only propagate the error with `?` (most places), or an enum specific to that method call when we need specific logic based on the error returned from the host (e.g. [here](https://github.com/cosmos/ibc-rs/blob/a39cfdec375e624d2ff75aeaee7556ae97d7beae/crates/ibc/src/core/ics04_channel/handler/recv_packet.rs#L131)). **Note that we currently use the same error types for [our handlers](https://github.com/cosmos/ibc-rs/blob/42b54613ba5c81b49557a8be46af517eec0b2fce/crates/ibc/src/core/ics02_client/handler/update_client.rs#L32) and for [host functions](https://github.com/cosmos/ibc-rs/blob/a39cfdec375e624d2ff75aeaee7556ae97d7beae/crates/ibc/src/core/context.rs#L81); I believe they should be distinct**.
+ The whole point of defining an `Error` enum instead of using a catch-all solution such as anyhow or eyre is to allow the user to write specific logic based on which error occurred. Hence, only the `Error` returned by our 2 entrypoints, `ValidationContext::validate()` and `ExecutionContext::execute()`, matters that it's an enum (even that is debatable), so that users can `match` against the return of such call.

Errors hold 2 purposes: (1) enabling users to write logic based on which error occured, and (2) providing a detailed human-readable report of an error stack (such as provided by eyre and anyhow). Whenever we find ourselves writing an enum, we're addressing (1), and we need to think carefully about how granular we want the error variants to be. For example, if the error returned by `validate()` looks like:

```rust
enum Error {
    ClientError,
    ConnectionError,
    ChannelError,
    PacketError,
}
```
then IMO this is unlikely that any user will be able to write any logic based on that (we can tell ahead of time which variant `validate()` will return based on the IBC message type), and we'd be better off using anyhow/eyre/snafu/whatever addresses (2).

TL;DR: I believe all internal errors should only focus on providing a context for top-level errors. `validate()` and `execute()` should return an enum if we can make it meaningful, or otherwise simply an error type that focuses on (2). Host functions (in `{Validation,Execution}Context`) should return errors distinct from internal library errors, and in almost all cases they're essentially simple `String`s focused on addressing (2).

**Comment by DaviRain-Su** [comment]
relate pr: https://github.com/cosmos/ibc-proto-rs/pull/38

File: crates/ibc/src/core/context.rs [hunk 4]
```diff
@@ -28,14 +28,39 @@ use super::{
...
         packet::{Receipt, Sequence},
         timeout::TimeoutHeight,
     },
     ics23_commitment::commitment::CommitmentPrefix,
     ics24_host::identifier::{ChannelId, ClientId, ConnectionId, PortId},
 };
+use displaydoc::Display;
+
+#[derive(Debug, Display)]
+pub enum ContextError {
```
Comment (DaviRain-Su) [inline crates/ibc/src/core/context.rs:42]: I do not know if my understanding of this is in line with the requirements

File: crates/ibc/Cargo.toml [hunk 5]
```diff
@@ -27,7 +44,7 @@ mocks = ["tendermint-testgen", "clock", "std"]
 
 [dependencies]
 # Proto definitions for all IBC-related interfaces, e.g., connections or channels.
-ibc-proto = { version = "0.22.0", default-features = false }
+ibc-proto = { version = "0.22.0", default-features = false, git = "https://github.com/octopus-network/ibc-proto-rs.git", branch = "refactor-error-system" }
```
Comment (plafer) [inline crates/ibc/Cargo.toml]: let's remember not to merge while we have this

File: crates/ibc/src/lib.rs [hunk 6]
```diff
@@ -4,6 +4,7 @@
 
 #![no_std]
 #![allow(clippy::large_enum_variant)]
+#![allow(clippy::result_large_err)]
```
Comment (plafer) [inline crates/ibc/src/lib.rs]: We should remove this [for performance reasons](https://rust-lang.github.io/rust-clippy/master/index.html#result_large_err); let's make our Error enums smaller instead.

File: crates/ibc/src/core/ics02_client/error.rs [hunk 7]
```diff
@@ -1,280 +1,186 @@
...
-            [ Ics23Error ]
-            | _ | { "ics23 verification failure" },
-
-        ClientSpecific
-            { description: String }
-            | e | { format_args!("client specific error: {0}", e.description) },
+#[derive(Debug, Display)]
+pub enum ClientError {
+    /// unknown client type: `{client_type}`
+    UnknownClientType { client_type: String },
```
Comment (plafer) [inline crates/ibc/src/core/ics02_client/error.rs]: Having thought about this some more, I like the direction this PR is taking. IMO what's left to be done is:
1. Remove all unused errors such as this one. Once this is done, we can additionally look to combine redundant errors.
2. "Roll our own backtrace", as @hu55a1n1 [suggested](https://github.com/cosmos/ibc-rs/pull/255#discussion_r1030775147). We should make sure to use the same convention everywhere, as you started doing here: `<error message>: {0}` (for when the error argument is another `Error`). When `std::error::Error` makes it into `core`, we can start using `thiserror`.
3. Make sure there's a `Error::Other` variant for host methods to use. This PR is getting big enough, we can change host functions in another PR.

File: crates/ibc/src/core/ics04_channel/error.rs [hunk 8]
```diff
@@ -12,354 +12,242 @@ use crate::signer::SignerError;
...
-        AppModule
-            { description: String }
-            | e | {
-                format_args!(
-                    "application module error: {0}",
-                    e.description)
-            },
+#[derive(Debug, Display)]
+pub enum ChannelError {
+    /// ICS03 connection error
```
Comment (plafer) [inline crates/ibc/src/core/ics04_channel/error.rs]:
```suggestion
    /// connection error: {0}
```
As discussed in other comment. We should change the other variants analogously

**Review by plafer — CHANGES_REQUESTED** [review]

**Comment by plafer** [comment]
Requested additional input from @blasrodri; let's hold until then

File: crates/ibc/src/lib.rs (hunk 6 above)
Comment (DaviRain-Su) [reply crates/ibc/src/lib.rs]: done

File: crates/ibc/src/core/ics02_client/error.rs (hunk 7 above)
Comment (DaviRain-Su) [reply crates/ibc/src/core/ics02_client/error.rs]: > When `std::error::Error` makes it into core, we can start using `thiserror`
Currently `thiserror` can not be used in no_std, `thiserror::Error` I think the implementation of this macro can not be used in the current no_std environment, and the current use of `displaydoc` I think similar to `thiserror` except that there is no error backtrace, but here we have implemented `std::error::Error`.

```rust
#[cfg(feature = "std")]
impl std::error::Error for Error {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match &self {
            Error::InvalidHeader { error: e, .. } => Some(e),
            Error::InvalidTendermintTrustThreshold(e) => Some(e),
            Error::InvalidChainIdentifier(e) => Some(e),
            Error::InvalidChainId { error: e, .. } => Some(e),
            Error::InvalidRawHeader(e) => Some(e),
            Error::Decode(e) => Some(e),
            Error::TimestampOverflow(e) => Some(e),
            _ => None,
        }
    }
}
```

File: crates/ibc/Cargo.toml (hunk 5 above)
Comment (DaviRain-Su) [reply crates/ibc/Cargo.toml]: done

**Comment by blasrodri** [comment]
I like the idea of getting rid of `flex-error`. Plus, the naming of the errors has become more descriptive, and do not assume that the developer is so acquainted with the different `ics`. I'm not sure if `thiserror` is needed at all - would be nice to see a use case that's not currently covered with the current implementation. Seems like something that the library consumers would like to have. 

Overall, I think this PR reflects the goals of #164 .

File: crates/ibc/src/core/ics04_channel/error.rs [hunk 9]
```diff
@@ -12,354 +12,221 @@ use crate::signer::SignerError;
...
-
-        AppModule
-            { description: String }
-            | e | {
-                format_args!(
-                    "application module error: {0}",
-                    e.description)
-            },
+#[derive(Debug, Display)]
+pub enum ChannelError {
```
Comment (plafer) [inline crates/ibc/src/core/ics04_channel/error.rs:18]: Here are the error variants that we still have that are unused:
+ `InvalidVersion`
+ `InvalidPacket`
+ `MissingPacket`
+ `MissingVersion`
+ `MissingNextSendSeq`

File: crates/ibc/src/core/ics02_client/error.rs [hunk 10]
```diff
@@ -1,280 +1,140 @@
...
-
-        Ics23Verification
-            [ Ics23Error ]
-            | _ | { "ics23 verification failure" },
-
-        ClientSpecific
-            { description: String }
-            | e | { format_args!("client specific error: {0}", e.description) },
+#[derive(Debug, Display)]
+pub enum ClientError {
```
Comment (plafer) [inline crates/ibc/src/core/ics02_client/error.rs:14]: Here are the error variants that we still have that are unused:
+ `InvalidRawConsensusState`
+ `MissingHeight`

File: crates/ibc/src/core/ics03_connection/error.rs [hunk 11]
```diff
@@ -7,168 +7,88 @@ use crate::signer::SignerError;
...
-            [ client_error::Error ]
-            | e | {
-                format_args!("the consensus proof verification failed (height: {0})",
-                    e.height)
-            },
-
+use displaydoc::Display;
+
+#[derive(Debug, Display)]
+pub enum ConnectionError {
```
Comment (plafer) [inline crates/ibc/src/core/ics03_connection/error.rs:13]: Here are the error variants that we still have that are unused:
+ `FrozenClient`
+ `ImplementationSpecific`

File: crates/ibc/src/core/ics04_channel/error.rs [hunk 12]
```diff
@@ -12,354 +12,221 @@ use crate::signer::SignerError;
...
+    ChannelFeatureNotSuportedByConnection,
+    /// the channel end (`{port_id}`, `{channel_id}`) does not exist
+    ChannelNotFound {
+        port_id: PortId,
+        channel_id: ChannelId,
+    },
+    /// Verification fails for the packet with the sequence number `{sequence}`, error(`{ics02_error}`)
+    PacketVerificationFailed {
+        sequence: Sequence,
+        ics02_error: client_error::ClientError,
```
Comment (plafer) [inline crates/ibc/src/core/ics04_channel/error.rs]:
```suggestion
        client_error: client_error::ClientError,
```

File: crates/ibc/src/core/ics04_channel/error.rs [hunk 13]
```diff
@@ -12,354 +12,221 @@ use crate::signer::SignerError;
...
+            _ => None,
+        }
     }
 }
 
-impl Error {
-    pub fn chan_open_confirm_proof_verification(e: Error) -> Error {
-        e.add_trace(&"Handshake proof verification fails at ChannelOpenConfirm")
+#[cfg(feature = "std")]
+impl std::error::Error for ChannelError {
```
Comment (plafer) [inline crates/ibc/src/core/ics04_channel/error.rs:206]: nit: can we move this under the `ChannelError` definition?

File: crates/ibc/src/core/ics04_channel/error.rs [hunk 14]
```diff
@@ -12,354 +12,221 @@ use crate::signer::SignerError;
...
-        AppModule
-            { description: String }
-            | e | {
-                format_args!(
-                    "application module error: {0}",
-                    e.description)
-            },
+#[derive(Debug, Display)]
+pub enum ChannelError {
+    /// ICS03 connection error
```
Comment (plafer) [inline crates/ibc/src/core/ics04_channel/error.rs]:
```suggestion
    /// connection error: `{0}`
```

File: crates/ibc/src/core/ics04_channel/error.rs [hunk 15]
```diff
@@ -12,354 +12,221 @@ use crate::signer::SignerError;
...
+    Connection(connection_error::ConnectionError),
+    /// ICS05 port error
```
Comment (plafer) [inline crates/ibc/src/core/ics04_channel/error.rs]:
```suggestion
    /// port error: `{0}`
```

File: crates/ibc/src/core/ics04_channel/error.rs (hunk 12 above)
Comment (plafer) [reply crates/ibc/src/core/ics04_channel/error.rs]: (and change the docstring accordingly)

File: crates/ibc/src/core/ics04_channel/error.rs (hunk 15 above)
Comment (plafer) [reply crates/ibc/src/core/ics04_channel/error.rs]: ... as well as `Signer`, `VerifyChannelFailed`, `InvalidIntAsSequence`, `InvalidProof`, and `Identifier`

File: crates/ibc/src/core/ics04_channel/error.rs [hunk 16]
```diff
@@ -12,354 +12,221 @@ use crate::signer::SignerError;
...
+    ConnectionNotOpen { connection_id: ConnectionId },
+    /// Receipt for the packet `{sequence}` not found
+    PacketReceiptNotFound { sequence: Sequence },
+    /// The stored commitment of the packet `{sequence}` is incorrect
+    IncorrectPacketCommitment { sequence: Sequence },
+    /// implementation specific error
+    ImplementationSpecific,
+    /// Undefined counterparty connection for `{connection_id}`
+    UndefinedConnectionCounterparty { connection_id: ConnectionId },
+    /// invalid proof
```
Comment (plafer) [inline crates/ibc/src/core/ics04_channel/error.rs]:
```suggestion
    /// invalid proof: `{0}`
```

**Review by plafer — CHANGES_REQUESTED** [review]
Thank you for this amazing work; we're getting really close to the finish line! 😃

File: crates/ibc/src/core/ics02_client/error.rs (hunk 7 above)
Comment (DaviRain-Su) [reply crates/ibc/src/core/ics02_client/error.rs]: when `thiserror` support we just add `#derive[thiserror::Error]` can remove below this.
```rust
#[cfg(feature = "std")]
impl std::error::Error for Error {}
```

File: crates/ibc/src/core/ics02_client/error.rs (hunk 10 above)
Comment (DaviRain-Su) [reply crates/ibc/src/core/ics02_client/error.rs:14]: done

File: crates/ibc/src/core/ics04_channel/error.rs (hunk 9 above)
Comment (DaviRain-Su) [reply crates/ibc/src/core/ics04_channel/error.rs:18]: done

File: crates/ibc/src/core/ics03_connection/error.rs (hunk 11 above)
Comment (DaviRain-Su) [reply crates/ibc/src/core/ics03_connection/error.rs:13]: done

File: crates/ibc/src/core/ics04_channel/error.rs (hunk 12 above)
Comment (DaviRain-Su) [reply crates/ibc/src/core/ics04_channel/error.rs]: done

File: crates/ibc/src/core/ics04_channel/error.rs (hunk 16 above)
Comment (DaviRain-Su) [reply crates/ibc/src/core/ics04_channel/error.rs]: done

File: crates/ibc/src/core/ics04_channel/error.rs (hunk 14 above)
Comment (DaviRain-Su) [reply crates/ibc/src/core/ics04_channel/error.rs]: done

File: crates/ibc/src/core/ics04_channel/error.rs (hunk 15 above)
Comment (DaviRain-Su) [reply crates/ibc/src/core/ics04_channel/error.rs]: done

File: crates/ibc/src/core/ics04_channel/error.rs (hunk 13 above)
Comment (DaviRain-Su) [reply crates/ibc/src/core/ics04_channel/error.rs:206]: I searched and did not find a relevant place to use

File: .changelog/unreleased/enhancement/164-refactor-error-system.md [hunk 17]
```diff
@@ -0,0 +1 @@
+- Refactor error system ([#164](https://github.com/cosmos/ibc-rs/issues/164))
```
Comment (plafer) [inline .changelog/unreleased/enhancement/164-refactor-error-system.md]:
```suggestion
- Remove `flex-error` and remove unused error variants ([#164](https://github.com/cosmos/ibc-rs/issues/164))
```

Comment (DaviRain-Su) [reply .changelog/unreleased/enhancement/164-refactor-error-system.md]: done

**Review by plafer — APPROVED** [review]
Awesome work :)

**Comment by DaviRain-Su** [comment]
> Awesome work :)

thx guys
